### DISCUSSION TITLE
How do can I create a re-usable object in my test setup, or anywhere, that can be used with all tests?

### DISCUSSION THREAD

**Discussion by drFrankWhite** [seleniumbase/SeleniumBase#2625] [General]
How do I create an object to test in the setUp() or in a fixtured function that can be used in all of that class's test cases?

**Comment by mdmintz** [comment]
See [examples/boilerplates/base_test_case.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/boilerplates/base_test_case.py)

**Reply by drFrankWhite** [reply]
When I use the setUp() below, each test creates a new test_db.  What am I doing wrong?  I only want to create on test_db for all tests in `TestDatabases`.
```
class TestDatabases(DatabaseBase):

    def setUp(self, **kwargs):
        super(TestDatabases, self).setUp()
        self.api = SomethingAPI()
        self.login()

        """POST a test_db"""
        post_resp = self.api.post_test_db("pg")
        if post_resp.status_code == 201:
            self.test_db = post_resp.json()["database"]  <--- runs for each test
        else:
            raise requests.HTTPError(f"{post_resp.status_code}: {post_resp.text}")
```

**Reply by mdmintz** [reply]
Overriding `setUp()` and `tearDown()` is for having methods that run before & after **every** test.
For just once, see: https://github.com/seleniumbase/SeleniumBase/issues/1243#issuecomment-1058069684

**Reply by drFrankWhite** [reply]
Ok, I do execute w/ `--rs`, but I thought that only reused the browser session, not the other objects (like `test_db`).  I want to create `test_db` once, then run a bunch of tests against it, then delete it in the `teardown()`.  #1243 seems like its addressing ordering tests.

Also, thank you for SB and your continued interaction with your users.

**Reply by mdmintz** [reply]
Since you're reusing the browser session for each test, just put your single-run setup code in `test_aaa()` and put the teardown code in `test_zzz()` (runs alphabetically). The standard `setUp()` and `tearDown()` methods are for each test (not once).

**Reply by drFrankWhite** [reply]
But if I create my `test_db` in `test_aa()`, the `test_db` object isn't available for the other tests ... 
`E       AttributeError: 'TestDatabases' object has no attribute 'test_db'`

**Reply by mdmintz** [reply]
Use `@classmethod` `def setUpClass(self):` and `@classmethod` `def tearDownClass(self):` for that then.

**Reply by drFrankWhite** [reply]
thank you!  this approach worked:
```
class TestDatabases(DatabaseBase):

    def setUp(self, **kwargs):
        super().setUp()
        self.login()

    @classmethod
    def setUpClass(cls):
        super(TestDatabases, cls).setUpClass()
        cls.api = DatabasesAPI()
        post_resp = cls.api.post_test_db("sql")
        if post_resp.status_code == 201:
            cls.test_db = post_resp.json()["database"]
        else:
            raise requests.HTTPError(f"{post_resp.status_code}: {post_resp.text}")

    @classmethod
    def tearDownClass(cls):
        cls.api.delete_db(cls.test_db["id"])
        super(TestDatabases, cls).tearDownClass()
```
